src/Cake.GitHub/Releases/GitHubReleaseCreator.cs

```diff
@@ -0,0 +1,216 @@
+                }
+            }
+
+            //
+            // Create new release
+            //
+            _cakeLog.Verbose("Creating Release");
+            var newRelease = new NewRelease(tagName)
+            {
+                TargetCommitish = settings.TargetCommitish,
+                Name = settings.Name,
+                Body = settings.Body,
+                Draft = settings.Draft,
+                Prerelease = settings.Prerelease
+            };
+
+            var createdRelease = await _githubClient.Repository.Release.Create(owner, repository, newRelease);
+            _cakeLog.Debug($"Created release with id '{createdRelease.Id}'");
+
+            var result = GitHubRelease.FromRelease(createdRelease);
+
+            //
+            // Upload assets
+            //
+            if (settings.AssetsOrEmpty.Count > 0)
+            {
+                _cakeLog.Verbose("Uploading Release Assets");
+                foreach (var filePath in settings.AssetsOrEmpty)
+                {
+                    _cakeLog.Debug($"Uploading asset '{filePath}'");
+                    using var stream = _fileSystem.GetFile(filePath).OpenRead();
+                    var assetUpload = new ReleaseAssetUpload()
+                    {
+                        FileName = filePath.GetFilename().ToString(),
+                        ContentType = "application/octet-stream",
+                        RawData = stream
+                    };
+                    var asset = await _githubClient.Repository.Release.UploadAsset(createdRelease, assetUpload);
+                    result.Add(GitHubReleaseAsset.FromReleaseAsset(asset));
+                }
+            }
+
+            _cakeLog.Verbose(settings.AssetsOrEmpty.Count > 0
+                ? $"Successfully created new GitHub Release and uploaded '{settings.AssetsOrEmpty.Count}' assets"
+                : "Successfully created new GitHub Release");
+
+            return result;
+        }
+
+        private async Task<Release?> TryGetReleaseAsync(string owner, string repository, string tagName)
+        {
+            try
+            {
+                var release = await _githubClient.Repository.Release.Get(owner, repository, tagName);
+                _cakeLog.Verbose($"Found existing release for tag '{tagName}'");
+                return release;
+            }
+            catch (NotFoundException)
+            {
+                // in case a release is a draft release, the tag has not been created yet and the release cannot be found via Get()
+                // to retrieve it, get all releases and search for the tag name                
+                var allReleases = await _githubClient.Repository.Release.GetAll(owner, repository);
+                var matchingReleases = allReleases.Where(x => StringComparer.Ordinal.Equals(x.TagName, tagName)).ToArray();
+
+                Release? release;
+                if(matchingReleases.Length == 0)
```

Comment:
can be replaced with SingleOrDefault()?

Reply:
I intentionally did not use `SingleOrDefault()` to get a better exception message rather than the generic message given by `SingleOrDefault()` and catching the exception just to rethrow it seemed wrong as well

Reply:
Makes sense, thanks for explaining.